feat: support fmt in media captions

dimpurr opened the pull request:
According to Issue https://github.com/grammyjs/parse-mode/issues/22 , this PR adds support for using fmt in media captions, including:

- `replyFmtWithPhoto`: send a photo with a formatted caption
- `replyFmtWithMediaGroup`: send a media group with formatted captions
- `editFmtMessageMedia`: edit media and formatted caption in a message
- `editFmtMessageText`: edit formatted text in a message

Changes made in this PR:

1. `deps.deno.ts` and `deps.node.ts`:
   - Import `InputMedia*` types from the correct paths.
   - Export `InputMedia*` types.

2. `hydrate.ts`:
   - Add new methods to `ParseModeFlavor` type: `replyFmtWithPhoto`, `replyFmtWithMediaGroup`, `editFmtMessageMedia`, `editFmtMessageText`.
   - Implement the new methods in `hydrateReplyFmt` middleware.
   - Update `replyFmtWithMediaGroup` to accept any `InputMedia*` type.

3. `README.md`:
   - Add documentation for the new methods in the "Using fmt in media captions" section.
   - Provide usage examples for the new methods.
   - Explain that `editFmtMessageMedia` allows changing the media type when editing a message.

Test: `deno run --allow-net --allow-read --allow-env test.ts`

```ts
import { Bot, Context } from "https://deno.land/x/grammy@v1.12.0/mod.ts";
import {
    hydrateReply,
    hydrateReplyFmt,
    ParseModeFlavor,
    bold,
    code,
    fmt,
    italic,
} from "./mod.ts";

const BOT_TOKEN = "";

const bot = new Bot<ParseModeFlavor<Context>>(BOT_TOKEN);

bot.use(hydrateReply);
bot.use(hydrateReplyFmt);

bot.command("start", async (ctx) => {
    await ctx.reply("Bot started. Send /test to test the plugin.");
});

bot.command("test", async (ctx) => {
    await ctx.replyFmt(fmt`This is a ${bold("formatted")} message.`);

    await ctx.replyFmtWithPhoto("http://placehold.co/600x400/EEE/31343C.jpg", {
        caption: fmt`This is a ${bold("formatted")} caption`,
    });

    await ctx.replyFmtWithMediaGroup([
        {
            type: "photo",
            media: "http://placehold.co/600x400/CCC/31343C.jpg",
            caption: fmt`First photo with ${italic("italic")} text`
        },
        {
            type: "video",
            media: "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerEscapes.mp4",
            caption: fmt`A ${code("video")} with formatted caption`
        }
    ]);
});

bot.catch((err) => {
    console.error(err);
});

bot.start();
```

KnightNiwrem commented on src/hydrate.ts:
```diff
@@ -18,6 +25,32 @@ type ParseModeFlavor<C extends Context> = C & {
+  ) => ReturnType<C["replyWithPhoto"]>;
+
+  replyFmtWithMediaGroup: (
+    media: (InputMediaPhoto | InputMediaVideo | InputMediaAnimation | InputMediaAudio | InputMediaDocument & { caption: Stringable })[]
```
The `caption` is always optional. We should not make it required here.

The manual union is workable. Pulling the type from `Parameters<C[themethodname]>` would be superior.

KnightNiwrem commented on src/hydrate.ts:
```diff
@@ -67,8 +100,72 @@ const middleware = async <C extends Context>(
   await next();
 };
 
+const hydrateReplyFmt = async <C extends Context>(
```
You can add these to the existing middleware. There is no compelling reasons to make 2 middlewares here.

dimpurr commented:
@KnightNiwrem Thank you for the review! Changes made:

- Merged hydrateReplyFmt into hydrateReply to avoid creating two separate middlewares.
- Fixed the type definitions for args in replyFmtWithPhoto and replyFmtWithMediaGroup using the Tail utility type.
- Ensured the correct passing of remaining arguments in replyFmtWithPhoto.

Updated test code:

```typescript
import { Bot, Context } from "https://deno.land/x/grammy@v1.12.0/mod.ts";
import {
  hydrateReply,
  ParseModeFlavor,
  bold,
  code,
  fmt,
  italic,
} from "./mod.ts";

const BOT_TOKEN = "";
const bot = new Bot<ParseModeFlavor<Context>>(BOT_TOKEN);

bot.use(hydrateReply);

bot.command("start", async (ctx) => {
  await ctx.reply("Bot started. Send /test to test the plugin.");
});

bot.command("test", async (ctx) => {
  await ctx.replyFmt(fmt`This is a ${bold("formatted")} message.`);

  await ctx.replyFmtWithPhoto("http://placehold.co/600x400/EEE/31343C.jpg", {
    caption: fmt`This is a ${bold("formatted")} caption`,
  });

  await ctx.replyFmtWithMediaGroup([
    {
      type: "photo",
      media: "http://placehold.co/600x400/CCC/31343C.jpg",
      caption: fmt`First photo with ${italic("italic")} text`,
    },
    {
      type: "video",
      media: "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerEscapes.mp4",
      caption: fmt`A ${code("video")} with formatted caption`,
    },
  ]);
});

bot.catch((err) => {
  console.error(err);
});

bot.start();
```

Test command:

```typescript
deno run --allow-net --allow-read --allow-env test.ts
```

Please let me know if any further suggestions or concerns.

KnightNiwrem commented:
Thanks for the PQ! I'll make some small changes as well.
